gridstatus/pjm.py

```diff
@@ -532,6 +624,450 @@ def _get_key(self):
+            lambda location_name: location_types.get(location_name, pd.NA),
+        )
+
+        return df
+
+    def _parse_lmp_item(self, item):
+        item_id = item["id"]
+        if item_id.endswith(" (DA)"):
+            market = Markets.DAY_AHEAD_HOURLY
+            item_id = item_id.replace(" (DA)", "")
+        else:
+            market = Markets.REAL_TIME_5_MIN
+        item_data = item["data"]
+        df = pd.DataFrame(item_data)
+        df["Market"] = market.value
+        df["Time"] = (
+            pd.to_datetime(df["timestamp"], unit="ms")
+            .dt.tz_localize(tz="UTC")
+            .dt.tz_convert(tz=self.default_timezone)
+        )
+        df["Location Name"] = item_id
+        return df
+
+    def __debug_nested_data(self, data):
+        pd.options.display.width = 0  # DEBUG
+        pairs = []
+        for k1, v1 in data.items():
+            if isinstance(v1, dict):
+                for k2, v2 in v1.items():
+                    if isinstance(v2, list) and len(v2) > 0:
+                        pairs.append(
+                            (
+                                k1,
+                                k2,
+                            ),
+                        )
+
+        for k1, k2 in pairs:
+            df = self._parse_lmp_series(data[k1][k2])
+            print(f"self._parse_lmp_series(data['{k1}']['{k2}'])")
+            print(df)
+
+    @staticmethod
+    def _dv_lmp_extract_chart_ids(response, verbose=False):
+        html = response.content
+        doc = bs4.BeautifulSoup(html, "html.parser")
+        scripts = doc.find_all("script", {"nonce": True})
+
+        chart_source_id = None
+        menu_form = doc.find("form", {"id": "menuForm"})
+        grandparent = menu_form.parent.parent
+        hidden_elements = grandparent.find_all("input", {"type": "hidden"})
+        j_hidden_elements = [
+            elem for elem in hidden_elements if elem["name"].startswith("j_")
+        ]
+        seed = j_hidden_elements[0]["name"]
+        for script in scripts:
+            lines = script.text.split("\n")
+            for line in lines:
+                match = re.search(rf's:"({seed}[^"]+)"', line)
+                if match is not None:
+                    chart_source_id = match.group(1)
+                    chart_parent_source_id = chart_source_id.split(":")[0]
+
+        if verbose:
+            print(f"chart_source_id = {chart_source_id}", file=sys.stderr)
+            print(f"chart_parent_source_id = {chart_parent_source_id}", file=sys.stderr)
+
+        if chart_source_id is not None and chart_parent_source_id is not None:
+            return {
+                "chart_source_id": chart_source_id,
+                "chart_parent_source_id": chart_parent_source_id,
+            }
+
+    @staticmethod
+    def _df_deduplicate(dfs, unique_cols, keep_field, keep_value, verbose=False):
+        """Deduplicate a list of dataframes based on a list of columns,
```

Comment:
what if you just sort the combined df by keep_field and keep_value, then do drop_duplicates(first/last). because you sorted by keep_value, that source will always get selected

Reply:
That was my original approach but then it was kinda ugly calculating whether to keep first or last depending on the alphabetical order of the _src values, i.e. dv and json. I don't feel great about reinventing this wheel either so I'll revisit this.